sdk/core/Azure.Core/src/Shared/SyncAsyncEventHandlerExtensions.cs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    /// <summary>
+    /// Extensions for raising <see cref="SyncAsyncEventHandler{T}"/>
+    /// events.
+    /// </summary>
+    internal static class SyncAsyncEventHandlerExtensions
+    {
+        /// <summary>
+        /// Raise an <see cref="Azure.Core.SyncAsyncEventHandler{T}"/>
+        /// event by executing each of the handlers sequentially (to avoid
+        /// introducing accidental parallelism in customer code) and collecting
+        /// any exceptions.
+        /// </summary>
+        /// <typeparam name="T">Type of the event arguments.</typeparam>
+        /// <param name="eventHandler">The event's delegate.</param>
+        /// <param name="e">
+        /// An <see cref="SyncAsyncEventArgs"/> instance that contains the
+        /// event data.
+        /// </param>
+        /// <param name="declaringTypeName">
+        /// The name of the type declaring the event to construct a helpful
+        /// exception message and distributed tracing span.
+        /// </param>
+        /// <param name="eventName">
+        /// The name of the event to construct a helpful exception message and
+        /// distributed tracing span.
+        /// </param>
+        /// <param name="clientDiagnostics">
+        /// Client diagnostics to wrap all the handlers in a new distributed
+        /// tracing span.
+        /// </param>
+        /// <returns>
+        /// A task that represents running all of the event's handlers.
+        /// </returns>
+        /// <exception cref="AggregateException">
+        /// An exception was thrown during the execution of at least one of the
+        /// event's handlers.
+        /// </exception>
+        public static async Task RaiseAsync<T>(
+            this SyncAsyncEventHandler<T> eventHandler,
+            T e,
+            string declaringTypeName,
+            string eventName,
+            ClientDiagnostics clientDiagnostics)
+            where T : SyncAsyncEventArgs
+        {
+            Argument.AssertNotNull(e, nameof(e));
+            Argument.AssertNotNullOrEmpty(declaringTypeName, nameof(declaringTypeName));
+            Argument.AssertNotNullOrEmpty(eventName, nameof(eventName));
+            Argument.AssertNotNull(clientDiagnostics, nameof(clientDiagnostics));
+
+            // Get the invocation list, but return early if there's no work
+            if (eventHandler == null) { return; }
+            Delegate[] handlers = eventHandler.GetInvocationList();
+            if (handlers == null || handlers.Length == 0) { return; }
+
+            // Wrap handler invocation in a distributed tracing span so it's
+            // easy for customers to track and measure
+            using DiagnosticScope scope = clientDiagnostics.CreateScope(declaringTypeName + "." + eventName);
```

Comment:
Should we just take the scope name?

Reply:
I use this for the exception message too, so I'd punt on that until there's a need.

Reply:
Yes, but in both cases it's `declaringTypeName + "." + eventName` that we can take as a single value.

Reply:
I still think it's worth naming the scopes for people so we don't end up with inconsistent names.  If it just says `operationName`, you'll get random values in there from new folks and we don't catch scope names in API reviews.  I added a variable to avoid doing the concat twice.